### PR TITLE
Step by step

### PR DESCRIPTION
**Pull request by idanarye**
@koreno - I want to use this for the installation tests. Instead of using a bare generator and hoping we counted the `yield`s and and `next` correctly and in sync, a `step_by_step` will allow us to enforce that the installation step the test thinks we are at is the same step the infra says we are at.

### CODE REVIEW

File: easypy/step_by_step.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        if self._current_step != step:
+            raise StepByStepSyncMismatch(step_by_step_name=self._name,
+                                         expected_step=step,
+                                         actual_step=self._current_step)
+
+    def __str__(self):
+        return '%s<%s>' % (self._name, self._current_step)
+
+
+def step_by_step(generator_func):
```
Comment (koreno) [inline easypy/step_by_step.py]: `lockstep`?

File: easypy/step_by_step.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            step_not_taken = next(invocation._generator)
+        except StopIteration:
+            # all is well - all steps were exhausted
+            invocation._current_step = 'finished'
+        else:
+            raise StepByStepSyncMismatch(step_by_step_name=invocation._name,
+                                         expected_step='finished',
+                                         actual_step=step_not_taken)
+
+    return wrapper
```
Comment (koreno) [inline easypy/step_by_step.py]: `wrapper.execute` -> executes all steps

File: easypy/lockstep.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        else:
+            raise LockstepSyncMismatch(lockstep_name=invocation._name,
+                                       expected_step='finished',
+                                       actual_step=step_not_taken)
+
+    @wraps(generator_func)
+    def wrapper(*args, **kwargs):
+        context = cm(*args, **kwargs)
+
+        def execute():
```
Comment (idanarye) [inline easypy/lockstep.py]: @koreno: This is a bit hacky, but I had to do it to implement the `execute` method as an attribute of the context object. I'm not sure we'll use it though - I think it would be better to create wrapper methods:

```python
@lockstep
def clusterize_lockstep(self, ...):
    # ...

def clusterize(self, ...):
    with self.clusterize_lockstep(...) as process:
        process.exhaust()
```

I put this feature is in a separate commit so that we can easily remove it if we want.

File: tests/test_lockstep.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        nonlocal finished
+
+        yield 'STEP_1'
+        yield 'STEP_2'
+        yield 'STEP_3'
+
+        finished = True
+
+    assert not finished
+    process().execute()
```
Comment (koreno) [inline tests/test_lockstep.py]: what I was thinking is `process.execute()`

File: easypy/lockstep.py (hunk 3 above)
Comment (koreno) [inline easypy/lockstep.py]: what I was thinking is `process.execute()`, not `process().execute()`, which I think simplifies.

not sure what you meant with "create wrapper methods"

File: easypy/lockstep.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            invocation._current_step = 'finished'
+        else:
+            raise LockstepSyncMismatch(lockstep_name=invocation._name,
+                                       expected_step='finished',
+                                       actual_step=step_not_taken)
+
+    @wraps(generator_func)
+    def wrapper(*args, **kwargs):
+        result = cm(*args, **kwargs)
+        result.__class__ = _LockstepGeneratorContextManager
```
Comment (idanarye) [inline easypy/lockstep.py]: @koreno - sorry for this ugly hack, but...

Somewhere between Python 3.4.2 and 3.4.5 they changed `_GeneratorContextManager.__init__` from
```python
class _GeneratorContextManager(ContextDecorator):
    """Helper for @contextmanager decorator."""

    def __init__(self, func, *args, **kwds):
```

to
```python
class _GeneratorContextManager(ContextDecorator):
    """Helper for @contextmanager decorator."""

    def __init__(self, func, args, kwds):
```

This means we can't use the same code to initiate both versions. Well, it was a private interface to begin with...

So instead, I'll use `contextmanager` to create the object, and switch it's class to something that inherits what `contextmanager` usually returns.

Comment (koreno) [inline easypy/lockstep.py]: must we switch the class? isn't it more polite to wrap it?

Comment (idanarye) [inline easypy/lockstep.py]: Can do
